File: bakamc-folia/src/main/kotlin/cn/bakamc/folia/flight_energy/FlightEnergyManager.kt
package cn.bakamc.folia.flight_energy

import cn.bakamc.common.text.bakatext.BakaText
import cn.bakamc.folia.config.FlightEnergyConfig.ALLOW_FLY_WORLD
import cn.bakamc.folia.config.FlightEnergyConfig.CLOSE_ADVENTURE_PLAYERS_FLYING
import cn.bakamc.folia.config.FlightEnergyConfig.ENERGY_COST
import cn.bakamc.folia.config.FlightEnergyConfig.FORBID_FLY_WORLD_MESSAGE
import cn.bakamc.folia.config.FlightEnergyConfig.MONEY_ITEM
import cn.bakamc.folia.config.FlightEnergyConfig.SYNC_PERIOD
import cn.bakamc.folia.config.FlightEnergyConfig.TICK_PERIOD
import cn.bakamc.folia.db.table.FlightEnergy
import cn.bakamc.folia.db.table.SpecialItem
import cn.bakamc.folia.extension.onlinePlayers
import cn.bakamc.folia.extension.uuid
import cn.bakamc.folia.item.SpecialItemManager
import cn.bakamc.folia.service.PlayerService
import cn.bakamc.folia.util.*
import kotlinx.coroutines.runBlocking
import moe.forpleuvoir.nebula.common.api.Initializable
import moe.forpleuvoir.nebula.common.color.Colors
import net.minecraft.server.level.ServerPlayer
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.jetbrains.annotations.Contract
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds
import kotlin.time.measureTime
import kotlin.time.measureTimedValue

object FlightEnergyManager : Listener, Initializable {

    private lateinit var energyCache: MutableMap<Player, FlightEnergy>

    private lateinit var energyBarCache: MutableMap<Player, EnergyBar>

    private lateinit var tasks: List<AsyncTask>

    private var syncing = AtomicBoolean(false)

    val Player.flightEnergy: FlightEnergy?
        get() = energyCache[this]

    val Player.energyBar: EnergyBar?
        get() = energyBarCache[this]

    /**
     * 玩家当前的飞行能量
     */
    var Player.energy: Double
        get() {
            return energyCache[this]?.energy ?: runBlocking { PlayerService.getFlightEnergy(this@energy).energy }
        }
        set(value) {
            energyCache[this]?.energy = value.coerceAtLeast(0.0)
            if (value <= 0.0) {
                toggleFly(this, false)
            }
        }

    var Player.barVisible: Boolean
        get() {
            return energyCache[this]?.barVisible ?: runBlocking { PlayerService.getFlightEnergy(this@barVisible).barVisible }
        }
        set(value) {
            energyCache[this]?.barVisible = value
        }

    var ServerPlayer.energy: Double
        get() {
            return energyCache.keys.find {
                it.uuid == this.stringUUID
            }?.energy ?: runBlocking {
                bakamc.server.getPlayer(this@energy.stringUUID)?.let {
                    PlayerService.getFlightEnergy(it).energy
                } ?: 0.0
            }

        }
        set(value) {
            energyCache.keys.find {
                it.uuid == this.stringUUID
            }?.let {
                energyCache[it]?.energy = value
            }
        }

    override fun init() {
        syncing.set(false)
        tasks = listOf(
            //tick
            AsyncTask(0.seconds, TICK_PERIOD) { tick() },
            //sync
            AsyncTask(1.minutes, SYNC_PERIOD) { sync() }
        )

        tasks.forEach { runAtFixedRate(it) }

        energyCache = ConcurrentHashMap()

        energyBarCache = ConcurrentHashMap()

        runBlocking {
            energyCache.putAll(PlayerService.getFlightEnergies(onlinePlayers))
            energyCache.forEach { (player, flightEnergy) ->
                energyBarCache[player] = EnergyBar.create(player, flightEnergy)
            }
            logger.info("飞行能量加载完成")
        }

    }

    fun onDisable() {
        if (this::energyCache.isInitialized) {
            if (!syncing.get()) sync()
            energyCache.clear()
            energyBarCache.forEach { (player, bar) ->
                bar.close()
            }
            energyBarCache.clear()
        }
    }

    /**
     * 获取货币对应的飞行能量
     * @return Map<SpecialItem, Double>
     */
    fun moneyItem(): Map<SpecialItem, Double> {
        return buildMap {
            SpecialItemManager.specifyType(MONEY_ITEM.keys).forEach { (key, specialItem) ->
                this[specialItem] = MONEY_ITEM[key]!!
            }
        }
    }

    /**
     * 获取指定货币对应的飞行能量
     * @param key String
     * @return Pair<SpecialItem, Double>?
     */
    fun moneyItem(key: String): Pair<SpecialItem, Double>? {
        return SpecialItemManager.specifyType(MONEY_ITEM.keys)[key]?.let {
            it to MONEY_ITEM[key]!!
        }
    }

    suspend fun onPlayerJoin(player: Player) {
        energyCache[player] = PlayerService.getFlightEnergy(player)
        energyBarCache[player] = EnergyBar.create(player, energyCache[player]!!)

        if (player.allowFly()) {
            player.flightEnergy?.let {
                player.allowFlight = it.enabled
                if (!player.isOnGround) player.isFlying = it.enabled
            }
        } else if (!player.inAllowFlyWorld) player.sendMessage(forbidFlyWorldMessage)

    }

    suspend fun onPlayerQuit(player: Player) {
        PlayerService.updateFlightEnergy(energyCache[player]!!)
        energyCache.remove(player)
        energyBarCache.remove(player)?.close()
    }

    fun onWorldChanged(event: PlayerChangedWorldEvent) {
        val player = event.player
        if (!player.allowFly()) {
            if (!player.inAllowFlyWorld) player.sendMessage(forbidFlyWorldMessage)
            toggleFly(player, false)
        }
    }

    fun onPlayerRespawn(player: Player) {
        if (player.allowFly()) {
            player.flightEnergy?.let {
                if (it.enabled) {
                    player.allowFlight = true
                }
            }
        } else {
            if (!player.inAllowFlyWorld) player.sendMessage(forbidFlyWorldMessage)
        }
    }

    fun onPlayerGameModeChange(player: Player, newGameMode: GameMode) {
        if (newGameMode == GameMode.SURVIVAL) {
            val isFlying = player.isFlying
            player.execute(1) {
                player.allowFlight = player.flightEnergy?.enabled ?: false
                if (player.allowFlight) player.isFlying = isFlying
            }
        } else {
            player.energyBar?.setVisible(false)
        }
    }

    fun addAllOnlinePlayerEnergy(energy: Double, range: ClosedRange<Double>): Int {
        energyCache.values.forEach {
            it.energy = (it.energy + energy).coerceIn(range)
        }
        if (syncing.get()) {
            runDelayed(1.seconds) {
                sync()
            }
        }
        return energyCache.size
    }

    /**
     * 切换飞行状态
     * @param player Player
     * @param enabled Boolean?
     */
    @Contract("_, null -> !enabled")
    fun toggleFly(player: Player, enabled: Boolean? = null) {
        player.flightEnergy?.let {
            if (enabled ?: !it.enabled)
                player.enableFly()
            else
                player.disableFly()
        }
    }

    private fun Player.enableFly() {
        if (!allowFly()) {
            if (!inAllowFlyWorld) sendMessage(forbidFlyWorldMessage)
            return
        }
        val flightEnergy = flightEnergy
        flightEnergy?.enabled = true
        if (flightEnergy?.enabled != true) energyBar?.setVisible(false)
        if (gameMode in arrayOf(GameMode.SURVIVAL, GameMode.ADVENTURE)) {
            allowFlight = flightEnergy?.enabled ?: false
        }
    }

    private fun Player.disableFly() {
        flightEnergy?.enabled = false
        energyBar?.setVisible(false)
        if (gameMode in arrayOf(GameMode.SURVIVAL, GameMode.ADVENTURE)) {
            allowFlight = false
            isFlying = false
        }
    }

    private val forbidFlyWorldMessage get() = BakaText.parse(FORBID_FLY_WORLD_MESSAGE)

    /**
     * 检查玩家是否可以被设置为飞行
     * @receiver Player
     * @return Boolean
     */
    private fun Player.allowFly(): Boolean {
        return inAllowFlyWorld && this.gameMode == GameMode.SURVIVAL && energy > 0
    }

    private val Player.inAllowFlyWorld: Boolean get() = world.name in ALLOW_FLY_WORLD

    private fun sync() {
        runBlocking {
            syncing.set(true)
            measureTimedValue {
                PlayerService.updateFlightEnergies(energyCache.values)
            }.let {
                logger.info("同步飞行能量成功,${it.value}条数据已更新，耗时${it.duration}")
            }
            syncing.set(false)
        }
    }

    suspend fun Player.updateEnergy(energy: Double) {
        val old = this.energy
        this.energy = energy
        measureTime {
            PlayerService.updateFlightEnergy(energyCache[this]!!)
        }.let {
            logger.info("玩家[${name}]飞行能量更新[${old} -> ${this.energy}],耗时$it")
        }
    }

    /**
     * 由单独线程控制循环
     * 每秒执行一次
     */
    private fun tick() {
        if (CLOSE_ADVENTURE_PLAYERS_FLYING) {
            onlinePlayers.filter { player -> player.gameMode == GameMode.ADVENTURE && player.flightEnergy!!.enabled }.forEach { player ->
                player.execute {
                    player.flightEnergy!!.enabled = false
                    player.allowFlight = false
                }
            }
        }
        onlinePlayers.filter { player ->
            //是否需要扣除飞行能量
            (player.gameMode in arrayOf(GameMode.SURVIVAL, GameMode.ADVENTURE)) && player.energy > 0.0 && player.vehicle == null
        }.filter { player ->
            //是否在可以飞行的世界且已经开了飞行
            if (!player.inAllowFlyWorld && player.flightEnergy!!.enabled) {
                player.disableFly()
                player.sendMessage(forbidFlyWorldMessage)
                return@filter false
            }
            //将飞行能量条可见设置为当前飞行状态
            player.energyBar?.setVisible(player.isFlying)
            player.isFlying
        }.forEach { player ->
            runCatching {
                //更新飞行能量
                player.energy = (player.energy - (ENERGY_COST)).coerceAtLeast(0.0)
                //更新飞行能量条状态
                player.energyBar?.tick()
                //如果飞行能量耗尽
                if (player.energy <= 0.0) {
                    //关闭飞行
                    toggleFly(player, false)
                    player.sendMessage(literalText("飞行能量已耗尽", Style(Colors.RED)))
                    //给予200tick的缓降效果
                    player.execute {
                        player.addPotionEffect(PotionEffect(PotionEffectType.SLOW_FALLING, 200, 1, false, true))
                    }

                }
            }.onFailure {
                logger.warn("玩家[${player.name}]飞行能量更新失败", it)
            }
        }
    }


}

File: bakamc-folia/src/main/scala/cn/bakamc/folia/functional/flightenergy/FlightEnergyManager.scala
package cn.bakamc.folia.functional.flightenergy

import cn.bakamc.folia.api.Reloadable
import cn.bakamc.folia.config.FlightEnergyConfig.*
import cn.bakamc.folia.database.services.PlayerService
import cn.bakamc.folia.database.table.FlightEnergy
import cn.bakamc.folia.functional.flightenergy.EnergyBar as _EnergyBar
import cn.bakamc.folia.util.AsyncTask
import cn.bakamc.folia.{BakaMC, execute, onlinePlayers}
import moe.forpleuvoir.nebula.common.api.Initializable
import moe.forpleuvoir.nebula.common.util.measureTime
import net.kyori.adventure.sound.Sound
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextColor
import org.bukkit.entity.Player
import org.bukkit.potion.{PotionEffect, PotionEffectType}
import org.bukkit.{GameMode, Sound as Sounds}

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import scala.concurrent.Await
import scala.concurrent.duration.{Duration, DurationDouble}
import scala.jdk.CollectionConverters.CollectionHasAsScala

object FlightEnergyManager extends Initializable, Reloadable {

  private[flightenergy] given [T] => Conversion[T, Option[T]] = Some(_)

  private def logger = BakaMC.logger

  private[flightenergy] val energyCache = ConcurrentHashMap[Player, FlightEnergy]()

  private[flightenergy] val barCache = ConcurrentHashMap[Player, EnergyBar]()

  private val tasks = List(
    AsyncTask(0.second, tickPeriod) { it => tick() },
    AsyncTask(0.second, syncPeriod) { it => sync() }
  )

  private[flightenergy] val allowGameMode = List(GameMode.SURVIVAL, GameMode.ADVENTURE)
  private[flightenergy] val timeOutDuration: Duration = 10.seconds

  given Conversion[AtomicBoolean, Boolean] = _.get()

  private[flightenergy] val syncing = AtomicBoolean(false)

  export FlightEnergyPlayerOps._

  export FlightEnergyCommandOps._

  override def initialization(): Unit = {
    syncing.set(false)

    tasks.foreach(_.run())

    energyCache.clear()
    barCache.clear()

    try {
      val result = Await.result(PlayerService.getFlightEnergies(onlinePlayers), timeOutDuration)
      result.foreachEntry { (player, energy) =>
        energyCache.put(player, energy)
        barCache.put(player, _EnergyBar(player, energy))
      }
      logger.info(s"同步玩家飞行能量成功,在线玩家: ${onlinePlayers.size},同步玩家: ${result.size}")
    } catch {
      case e: Throwable => logger.error(s"同步玩家飞行能量失败: ${e.getMessage}")
    }
  }

  def onDisable(): Unit = {
    if (!syncing) sync()
    energyCache.clear()
    barCache.forEach((player, bar) => bar.close())
    barCache.clear()
    tasks.foreach(_.cancel())
  }

  override def reload(): Unit = {
    onDisable()
    initialization()
  }

  private[flightenergy] def sync(): Unit = {
    if (syncing) return
    syncing.set(true)
    measureTime {
      try Await.result(PlayerService.updateFlightEnergies(energyCache.values.asScala), timeOutDuration)
      catch {
        case e: Throwable => logger.error(s"玩家飞行能量同步失败: ${e.getMessage}")
          None
      } finally {
        syncing.set(false)
      }
    }.foreach { (result, time) =>
      if (result.exists(_ > 0)) {
        logger.info(s"玩家飞行能量同步成功,同步玩家: ${result.size},同步耗时: ${time.toMillis}ms")
      }
    }
  }

  private def tick(): Unit = {
    onlinePlayers.foreach { player =>
      val gameMode = player.getGameMode //玩家游戏模式
      val isFlying = player.isFlying //玩家是否正在飞行
      val isEnabled = player.flightEnergy.exists(_.enabled) //玩家是否开启了飞行

      //如果禁止冒险模式飞行&玩家处于冒险模式,并且玩家开启了插件的飞行开关,关闭玩家所有飞行
      if (disableAdventureModeFly && gameMode == GameMode.ADVENTURE && isEnabled) {
        player.execute() {
          player.flightEnergy.foreach(_.enabled = false) //关闭插件的飞行开关
          player.setAllowFlight(false) //关闭原版的飞行开关
        }
      }

      //是否在可以飞行的世界
      val inAllowFlyWorld = player.inAllowFlyWorld
      val shouldShowBar = allowGameMode.contains(gameMode) && // 是否处于冒险模式和生存模式
        player.getVehicle == null && //不能处于骑乘状态下
        player.energy > 0.0 && //有飞行能量
        inAllowFlyWorld //是否在可以飞行的世界

      //更新飞行能量条显示状态
      player.energyBar.foreach(_.visible = shouldShowBar && isFlying)

      if (shouldShowBar) {
        //是否在可以飞行的世界且已经开了飞行
        if (!inAllowFlyWorld && isEnabled) {
          //关闭飞行并提醒玩家
          player.disableFly()
          player.sendMessage(forbidFlyWorldMessageComponent)
        }
        //正在飞行 处理扣费
        else if (isFlying) {
          try {
            //扣除能量
            player.energy = player.energy - energyCost
            //更新能量条状态
            player.energyBar.foreach(_.tick())
            //玩家能量小于等于0
            if (player.energy <= 0.0) {
              //关闭飞行
              player.toggleFlyState(false)
              //提醒玩家
              player.sendMessage(Component.text("飞行能量已耗尽").color(TextColor.color(0xFF0000)))
              logger.info(s"玩家 ${player.getName} 的飞行能量已耗尽")
              //给玩家添加200tick的缓降效果
              player.execute() {
                player.playSound(Sound.sound(Sounds.BLOCK_ANVIL_LAND, Sound.Source.UI, 1.0f, 1.0f))
                player.addPotionEffect(new PotionEffect(PotionEffectType.SLOW_FALLING, 200, 0))
              }
            }
          } catch {
            case e: Throwable => logger.warn(s"玩家 ${player.getName} 飞行能量更新失败: ${e.getMessage}")
          }
        }
      }
    }

  }

}




